micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c

```diff
@@ -0,0 +1,325 @@
+    // encrypted counter
+    uint8_t encrypted_counter[16];
+};
+
+#ifdef HAVE_CONFIG_H
+    #include <config.h>
+#endif
+#ifndef WOLFSSL_USER_SETTINGS
+    #include <wolfssl/options.h>
+#endif
+#include <wolfssl/wolfcrypt/settings.h>
+
+#include <wolfssl/wolfcrypt/aes.h>
+
+struct wolfssl_aes_ctx_with_key {
+    union {
+        Aes ctx;
+        struct {
+            uint8_t key[32];
+            uint8_t keysize;
+        } init_data;
+    } u;
+    unsigned char iv[16];
+};
+#define AES_CTX_IMPL struct wolfssl_aes_ctx_with_key
+
+typedef struct _mp_obj_aes_t {
+    mp_obj_base_t base;
+    AES_CTX_IMPL ctx;
+    uint8_t block_mode : 6;
+#define AES_KEYTYPE_NONE 0
+#define AES_KEYTYPE_ENC  1
+#define AES_KEYTYPE_DEC  2
+    uint8_t key_type : 2;
+} mp_obj_aes_t;
+
+static inline bool is_ctr_mode(int block_mode) {
+    #if MICROPY_PY_UCRYPTOLIB_CTR
+    return block_mode == UCRYPTOLIB_MODE_CTR;
+    #else
+    return false;
+    #endif
+}
+
+static inline struct ctr_params *ctr_params_from_aes(mp_obj_aes_t *o) {
+    // ctr_params follows aes object struct
+    return (struct ctr_params *)&o[1];
+}
+
+STATIC void aes_initial_set_key_impl(AES_CTX_IMPL *ctx, const uint8_t *key, size_t keysize, const uint8_t iv[16]) {
+    ctx->u.init_data.keysize = keysize;
+    memcpy(ctx->u.init_data.key, key, keysize);
+
+    if (NULL != iv) {
+        memcpy(ctx->iv, iv, sizeof(ctx->iv));
+    }
+}
+
+STATIC void aes_final_set_key_impl(AES_CTX_IMPL *ctx, bool encrypt, mp_int_t block_mode) {
+    // first, copy key aside
+    uint8_t key[32];
+    uint8_t keysize = ctx->u.init_data.keysize;
+    memcpy(key, ctx->u.init_data.key, keysize);
+
+    assert(16 == keysize || 32 == keysize);
+    int dir = (encrypt) ? AES_ENCRYPTION : AES_DECRYPTION;
+
+    // now, override key with the context object, calling the appropriate key initialization
+    // function for the block mode
+    if (block_mode == UCRYPTOLIB_MODE_CBC) {
+        wc_AesSetKey(&ctx->u.ctx, key, keysize, (const unsigned char*)&ctx->iv, dir);
+    } else if (block_mode == UCRYPTOLIB_MODE_ECB) {
+        // ECB requires NULL IV
+        wc_AesSetKey(&ctx->u.ctx, key, keysize, NULL, dir);
+    } else if (block_mode == UCRYPTOLIB_MODE_CTR) {
+        // from AES API documentation:
+        // NOTE: If using wc_AesSetKeyDirect with Aes Counter mode (Stream cipher)
+        // only use AES_ENCRYPTION for both encrypting and decrypting
+        wc_AesSetKeyDirect(&ctx->u.ctx, key, keysize, (const unsigned char*)&ctx->iv, AES_ENCRYPTION);
+    }
+}
+
+STATIC void aes_process_ecb_impl(AES_CTX_IMPL *ctx, const uint8_t in[16], uint8_t out[16], bool encrypt) {
+    if (encrypt) {
+        wc_AesEncryptDirect(&ctx->u.ctx, out, in);
+    } else {
+        wc_AesDecryptDirect(&ctx->u.ctx, out, in);
+    }
+}
+
+STATIC void aes_process_cbc_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *out, size_t in_len, bool encrypt) {
+    if (encrypt) {
+        wc_AesCbcEncrypt(&ctx->u.ctx, out, in, in_len);
```

Comment:
Is there an error log we could print to in the case of these wolfSSL Aes Encrypt/Decrypt functions returning an error value?